File: mcp-server/handlers/memory.py
#!/usr/bin/env python3
"""
MemOS MCP Server Memory Handlers

Handles memos_save, memos_list, memos_list_v2, memos_get, memos_get_stats tools.

Phase 1 Enhancement: Context compression for efficient token usage.
Phase 1.1: Deduplication to prevent duplicate saves from hooks + manual calls.
"""

import hashlib
import time
from typing import Any

import httpx

from api_client import api_call_with_retry
from config import MEMOS_URL, MEMOS_USER, logger
from cube_manager import ensure_cube_registered
from formatters import format_memories_for_display
from mcp.types import TextContent
from models import (
    COMPACTION_THRESHOLD,
    PREVIEW_COUNT,
    CompactedSearchResult,
    should_compact,
    to_full,
    to_minimal,
)
from query_processing import compute_memory_stats

from handlers.utils import (
    ERR_PARAM_MISSING,
    api_error_response,
    cube_registration_error,
    error_response,
    get_cube_id_from_args,
)

# ============================================================================
# Deduplication Cache
# ============================================================================
# Prevents duplicate saves within a short time window (e.g., hook + manual save)

# Cache: {content_hash: (cube_id, timestamp)}
_save_dedup_cache: dict[str, tuple[str, float]] = {}
DEDUP_TTL_SECONDS = 60  # Ignore duplicate saves within 60 seconds


def _content_hash(content: str, cube_id: str) -> str:
    """Generate hash for content + cube_id combination."""
    key = f"{cube_id}:{content}"
    return hashlib.md5(key.encode()).hexdigest()


def _is_duplicate_save(content: str, cube_id: str) -> bool:
    """
    Check if this content was recently saved (within TTL).
    Returns True if duplicate, False if new.
    """
    content_key = _content_hash(content, cube_id)
    now = time.time()

    # Clean expired entries
    expired = [k for k, (_, ts) in _save_dedup_cache.items() if now - ts > DEDUP_TTL_SECONDS]
    for k in expired:
        del _save_dedup_cache[k]

    # Check if duplicate
    if content_key in _save_dedup_cache:
        _, saved_at = _save_dedup_cache[content_key]
        if now - saved_at < DEDUP_TTL_SECONDS:
            logger.debug(f"Duplicate save detected (within {DEDUP_TTL_SECONDS}s), skipping")
            return True

    return False


def _mark_saved(content: str, cube_id: str) -> None:
    """Mark content as saved in dedup cache."""
    content_key = _content_hash(content, cube_id)
    _save_dedup_cache[content_key] = (cube_id, time.time())


async def handle_memos_save(
    client: httpx.AsyncClient,
    arguments: dict[str, Any]
) -> list[TextContent]:
    """Handle memos_save tool call with deduplication."""
    cube_id = get_cube_id_from_args(arguments)
    content = arguments.get("content", "")
    memory_type = arguments.get("memory_type")

    # memory_type is now REQUIRED - reject if not provided
    if not memory_type:
        return error_response(
            "memory_type parameter is required",
            error_code=ERR_PARAM_MISSING,
            suggestions=[
                "Bug fix -> `BUGFIX` or `ERROR_PATTERN`",
                "Technical decision -> `DECISION`",
                "Gotcha/trap -> `GOTCHA`",
                "Code template -> `CODE_PATTERN`",
                "Config change -> `CONFIG`",
                "New feature -> `FEATURE`",
                "Major achievement -> `MILESTONE`",
                "Pure progress update -> `PROGRESS`",
                "Example: `memos_save(content=\"...\", memory_type=\"BUGFIX\")`",
            ],
        )

    # Prepend memory type if not already present
    if not content.startswith(f"[{memory_type}]"):
        content = f"[{memory_type}] {content}"

    # Deduplication check - prevent duplicate saves within TTL
    if _is_duplicate_save(content, cube_id):
        return [TextContent(
            type="text",
            text=f"⏭️ Skipped: Same content was saved within {DEDUP_TTL_SECONDS}s (dedup protection)"
        )]

    # Auto-register cube if needed
    reg_success, reg_error = await ensure_cube_registered(client, cube_id)
    if not reg_success:
        return cube_registration_error(cube_id, reg_error)

    success, data, status = await api_call_with_retry(
        client, "POST", f"{MEMOS_URL}/memories", cube_id,
        json={
            "user_id": MEMOS_USER,
            "mem_cube_id": cube_id,
            "memory_content": content
        }
    )

    if success:
        # Mark as saved in dedup cache
        _mark_saved(content, cube_id)
        return [TextContent(type="text", text=f"Memory saved as [{memory_type}] in cube '{cube_id}'")]
    elif data:
        return api_error_response("Save", data.get("message", "Unknown error"))
    else:
        return api_error_response("Save", f"HTTP {status}")


async def handle_memos_list(
    client: httpx.AsyncClient,
    arguments: dict[str, Any]
) -> list[TextContent]:
    """Handle memos_list and memos_list_v2 tool calls with context compression."""
    cube_id = get_cube_id_from_args(arguments)
    limit = arguments.get("limit", 20)
    memory_type = arguments.get("memory_type")
    compact = arguments.get("compact", True)  # Default to compact mode

    # Auto-register cube
    reg_success, reg_error = await ensure_cube_registered(client, cube_id)
    if not reg_success:
        return cube_registration_error(cube_id, reg_error)

    params = {
        "user_id": MEMOS_USER,
        "mem_cube_id": cube_id,
        "limit": limit
    }
    if memory_type:
        params["memory_type"] = memory_type

    success, data, status = await api_call_with_retry(
        client, "GET", f"{MEMOS_URL}/memories", cube_id,
        params=params
    )

    if success and data:
        result_data = data.get("data", {})

        # Extract all memories for counting
        all_memories = _extract_memories_from_data(result_data)
        total_count = len(all_memories)

        # Apply context compression if enabled and threshold exceeded
        if compact and should_compact(total_count):
            logger.debug(f"Compacting list results: {total_count} > {COMPACTION_THRESHOLD}")
            preview = [to_minimal(m) for m in all_memories[:PREVIEW_COUNT]]
            compacted = CompactedSearchResult(
                preview=preview,
                total_count=total_count,
                omitted_count=total_count - len(preview),
                message="Use memos_get(memory_id=\"<id>\") for full details",
                query=f"list (type={memory_type})" if memory_type else "list all",
                cube_id=cube_id,
            )
            return [TextContent(type="text", text=compacted.to_display_text())]

        # Standard full display
        formatted = format_memories_for_display(result_data)
        return [TextContent(type="text", text=formatted)]
    elif data:
        return api_error_response("List", data.get("message", "Unknown error"))
    else:
        return api_error_response("List", f"HTTP {status}")


async def handle_memos_get_stats(
    client: httpx.AsyncClient,
    arguments: dict[str, Any]
) -> list[TextContent]:
    """Handle memos_get_stats tool call."""
    cube_id = get_cube_id_from_args(arguments)

    # Auto-register cube if needed
    reg_success, reg_error = await ensure_cube_registered(client, cube_id)
    if not reg_success:
        return cube_registration_error(cube_id, reg_error)

    success, data, status = await api_call_with_retry(
        client, "GET", f"{MEMOS_URL}/memories", cube_id,
        params={"user_id": MEMOS_USER, "mem_cube_id": cube_id}
    )

    if success and data:
        # Use helper function to compute stats
        stats, total = compute_memory_stats(data.get("data", {}))

        if not stats:
            return [TextContent(type="text", text=f"No memories found in cube '{cube_id}'.")]

        result = [f"## 📊 Memory Stats: {cube_id}"]
        result.append(f"Total Memories: **{total}**\n")
        for mtype, count in sorted(stats.items(), key=lambda x: x[1], reverse=True):
            percentage = (count / total) * 100
            result.append(f"- **{mtype}**: {count} ({percentage:.1f}%)")

        # Health check: High PROGRESS ratio warning
        progress_count = stats.get("PROGRESS", 0)
        if total > 0 and progress_count / total > 0.7:
            result.append("")
            result.append("---")
            result.append("")
            result.append(f"⚠️ **健康警告**: PROGRESS 类型占比过高 (>{progress_count / total:.0%})")
            result.append("")
            result.append("这可能导致 Neo4j 知识图谱无法建立有效关系。建议:")
            result.append("1. 保存记忆时**显式指定** `memory_type` 参数")
            result.append("2. 参考类型选择决策树:")
            result.append("   - 修复 Bug → `BUGFIX` 或 `ERROR_PATTERN`")
            result.append("   - 技术决策 → `DECISION`")
            result.append("   - 发现陷阱 → `GOTCHA`")
            result.append("   - 新功能 → `FEATURE`")
            result.append("   - 配置变更 → `CONFIG`")

        return [TextContent(type="text", text="\n".join(result))]
    elif data:
        return api_error_response("Stats", data.get("message", "Unknown error"))
    else:
        return api_error_response("Stats", f"HTTP {status}")


def _extract_memories_from_data(data: dict) -> list[dict]:
    """
    Extract flat list of memories from API response data.

    Handles both tree_text mode (nested nodes) and flat list responses.
    """
    memories = []

    text_mems = data.get("text_mem", [])
    for cube_data in text_mems:
        memories_data = cube_data.get("memories", [])

        # Handle tree_text mode (dict with nodes)
        if isinstance(memories_data, dict) and "nodes" in memories_data:
            memories.extend(memories_data["nodes"])
        elif isinstance(memories_data, list):
            memories.extend(memories_data)

    return memories


async def handle_memos_get(
    client: httpx.AsyncClient,
    arguments: dict[str, Any]
) -> list[TextContent]:
    """
    Handle memos_get tool call - get full details of a single memory by ID.

    This is the complement to compacted search/list results.
    When search returns a CompactedSearchResult, use this to get full details.
    """
    cube_id = get_cube_id_from_args(arguments)
    memory_id = arguments.get("memory_id", "")

    if not memory_id:
        return error_response(
            "memory_id parameter is required",
            error_code=ERR_PARAM_MISSING,
            suggestions=[
                "Get memory_id from memos_search or memos_list_v2 results",
                "Example: `memos_get(memory_id=\"abc123-...\")`",
            ],
        )

    # Auto-register cube if needed
    reg_success, reg_error = await ensure_cube_registered(client, cube_id)
    if not reg_success:
        return cube_registration_error(cube_id, reg_error)

    # Use direct API endpoint: GET /memories/{mem_cube_id}/{memory_id}
    success, data, status = await api_call_with_retry(
        client, "GET", f"{MEMOS_URL}/memories/{cube_id}/{memory_id}", cube_id
    )

    if success and data:
        result_data = data.get("data", {})

        if result_data:
            # Convert to full model and format
            full_mem = to_full(result_data, cube_id=cube_id, user_id=MEMOS_USER)

            lines = [
                f"## 📝 Memory Details",
                f"",
                f"**ID**: `{full_mem.id}`",
                f"**Type**: {full_mem.memory_type}",
                f"**Cube**: {full_mem.cube_id}",
            ]

            if full_mem.key:
                lines.append(f"**Key**: {full_mem.key}")
            if full_mem.tags:
                lines.append(f"**Tags**: {', '.join(full_mem.tags)}")
            if full_mem.created_at:
                lines.append(f"**Created**: {full_mem.created_at}")

            lines.append("")
            lines.append("### Content")
            lines.append("")
            lines.append(full_mem.content)

            if full_mem.background:
                lines.append("")
                lines.append("### Background")
                lines.append("")
                lines.append(full_mem.background)

            if full_mem.relations:
                lines.append("")
                lines.append("### Relations")
                lines.append("")
                for rel in full_mem.relations[:5]:  # Limit relations
                    lines.append(f"- {rel}")

            return [TextContent(type="text", text="\n".join(lines))]
        else:
            return [TextContent(
                type="text",
                text=f"❌ Memory not found: `{memory_id}`\n\n"
                     f"💡 **Tips**:\n"
                     f"- Verify the ID is correct (copy from memos_search results)\n"
                     f"- The memory may have been deleted\n"
                     f"- Try `memos_search` to find the memory again"
            )]
    elif data:
        error_msg = data.get("message", "Unknown error")
        if "not found" in error_msg.lower() or status == 404:
            return [TextContent(
                type="text",
                text=f"❌ Memory not found: `{memory_id}`\n\n"
                     f"💡 **Tips**:\n"
                     f"- Verify the ID is correct (copy from memos_search results)\n"
                     f"- The memory may have been deleted\n"
                     f"- Try `memos_search` to find the memory again"
            )]
        return api_error_response("Get", error_msg)
    else:
        return api_error_response("Get", f"HTTP {status}")

File: mcp-server-node/src/handlers/memory.ts
/**
 * Memory Handlers
 *
 * memos_save, memos_list_v2, memos_get, memos_get_stats
 */

import * as crypto from "crypto";
import { MEMOS_URL, MEMOS_USER, logger } from "../config.js";
import { apiCallWithRetry } from "../api-client.js";
import { ensureCubeRegistered } from "../cube-manager.js";
import { formatMemoriesForDisplay } from "../formatters.js";
import {
  COMPACTION_THRESHOLD,
  PREVIEW_COUNT,
  shouldCompact,
  compactedResultToText,
  toMinimal,
  toFull,
} from "../models.js";
import { computeMemoryStats, extractMcpType } from "../query-processing.js";
import type { TextContent, MemoryNode, SearchData } from "../types.js";
import {
  ERR_PARAM_MISSING,
  apiErrorResponse,
  cubeRegistrationError,
  errorResponse,
  getCubeIdFromArgs,
} from "./utils.js";

// ============================================================================
// Deduplication Cache
// ============================================================================

const saveDedupCache: Map<string, [string, number]> = new Map();
const DEDUP_TTL_SECONDS = 60;

function contentHash(content: string, cubeId: string): string {
  return crypto.createHash("md5").update(`${cubeId}:${content}`).digest("hex");
}

function isDuplicateSave(content: string, cubeId: string): boolean {
  const key = contentHash(content, cubeId);
  const now = Date.now() / 1000;

  // Clean expired
  for (const [k, [, ts]] of saveDedupCache) {
    if (now - ts > DEDUP_TTL_SECONDS) saveDedupCache.delete(k);
  }

  const entry = saveDedupCache.get(key);
  if (entry && now - entry[1] < DEDUP_TTL_SECONDS) {
    logger.debug(`Duplicate save detected (within ${DEDUP_TTL_SECONDS}s), skipping`);
    return true;
  }
  return false;
}

function markSaved(content: string, cubeId: string): void {
  const key = contentHash(content, cubeId);
  saveDedupCache.set(key, [cubeId, Date.now() / 1000]);
}

// ============================================================================
// memos_save
// ============================================================================

export async function handleMemosSave(arguments_: Record<string, unknown>): Promise<TextContent[]> {
  const cubeId = getCubeIdFromArgs(arguments_);
  let content = String(arguments_.content ?? "");
  const memoryType = arguments_.memory_type as string | undefined;

  if (!memoryType) {
    return errorResponse(
      "memory_type parameter is required",
      ERR_PARAM_MISSING,
      [
        "Bug fix -> `BUGFIX` or `ERROR_PATTERN`",
        "Technical decision -> `DECISION`",
        "Gotcha/trap -> `GOTCHA`",
        "Code template -> `CODE_PATTERN`",
        "Config change -> `CONFIG`",
        "New feature -> `FEATURE`",
        "Major achievement -> `MILESTONE`",
        "Pure progress update -> `PROGRESS`",
        'Example: `memos_save(content="...", memory_type="BUGFIX")`',
      ]
    );
  }

  if (!content.startsWith(`[${memoryType}]`)) {
    content = `[${memoryType}] ${content}`;
  }

  if (isDuplicateSave(content, cubeId)) {
    return [{ type: "text", text: `⏭️ Skipped: Same content was saved within ${DEDUP_TTL_SECONDS}s (dedup protection)` }];
  }

  const [regSuccess, regError] = await ensureCubeRegistered(cubeId);
  if (!regSuccess) return cubeRegistrationError(cubeId, regError);

  const result = await apiCallWithRetry(
    "POST",
    `${MEMOS_URL}/memories`,
    cubeId,
    {
      body: {
        user_id: MEMOS_USER,
        mem_cube_id: cubeId,
        memory_content: content,
      },
    },
    ensureCubeRegistered
  );

  if (result.success) {
    markSaved(content, cubeId);
    return [{ type: "text", text: `Memory saved as [${memoryType}] in cube '${cubeId}'` }];
  } else if (result.data) {
    return apiErrorResponse("Save", String((result.data as Record<string, unknown>).message ?? "Unknown error"));
  } else {
    return apiErrorResponse("Save", `HTTP ${result.status}`);
  }
}

// ============================================================================
// memos_list_v2
// ============================================================================

export async function handleMemosList(arguments_: Record<string, unknown>): Promise<TextContent[]> {
  const cubeId = getCubeIdFromArgs(arguments_);
  const limit = Number(arguments_.limit ?? 20);
  const memoryType = arguments_.memory_type as string | undefined;
  const compact = arguments_.compact !== false; // Default true

  const [regSuccess, regError] = await ensureCubeRegistered(cubeId);
  if (!regSuccess) return cubeRegistrationError(cubeId, regError);

  const params: Record<string, string | number | boolean> = {
    user_id: MEMOS_USER,
    mem_cube_id: cubeId,
  };
  if (!memoryType) {
    params.limit = limit;
  }

  const result = await apiCallWithRetry("GET", `${MEMOS_URL}/memories`, cubeId, { params }, ensureCubeRegistered);

  if (result.success && result.data) {
    const resultData = (result.data as Record<string, unknown>).data as SearchData ?? {};
    let allMemories = extractMemoriesFromData(resultData);

    if (memoryType) {
      allMemories = allMemories.filter((m) => extractMcpType(m) === memoryType).slice(0, limit);
    }

    const totalCount = allMemories.length;

    if (compact && shouldCompact(totalCount)) {
      const preview = allMemories.slice(0, PREVIEW_COUNT).map(toMinimal);
      const text = compactedResultToText({
        preview,
        totalCount,
        omittedCount: totalCount - preview.length,
        message: 'Use memos_get(memory_id="<id>") for full details',
        query: memoryType ? `list (type=${memoryType})` : "list all",
        cubeId,
      });
      return [{ type: "text", text }];
    }

    const formatted = formatMemoriesForDisplay(resultData);
    return [{ type: "text", text: formatted }];
  } else if (result.data) {
    return apiErrorResponse("List", String((result.data as Record<string, unknown>).message ?? "Unknown error"));
  } else {
    return apiErrorResponse("List", `HTTP ${result.status}`);
  }
}

// ============================================================================
// memos_get
// ============================================================================

export async function handleMemosGet(arguments_: Record<string, unknown>): Promise<TextContent[]> {
  const cubeId = getCubeIdFromArgs(arguments_);
  const memoryId = String(arguments_.memory_id ?? "");

  if (!memoryId) {
    return errorResponse(
      "memory_id parameter is required",
      ERR_PARAM_MISSING,
      [
        "Get memory_id from memos_search or memos_list_v2 results",
        'Example: `memos_get(memory_id="abc123-...")`',
      ]
    );
  }

  const [regSuccess, regError] = await ensureCubeRegistered(cubeId);
  if (!regSuccess) return cubeRegistrationError(cubeId, regError);

  const result = await apiCallWithRetry(
    "GET",
    `${MEMOS_URL}/memories/${cubeId}/${memoryId}`,
    cubeId,
    {},
    ensureCubeRegistered
  );

  if (result.success && result.data) {
    const resultData = (result.data as Record<string, unknown>).data as MemoryNode | null;

    if (resultData) {
      const fullMem = toFull(resultData, cubeId, MEMOS_USER);
      const lines = [
        "## 📝 Memory Details",
        "",
        `**ID**: \`${fullMem.id}\``,
        `**Type**: ${fullMem.memoryType}`,
        `**Cube**: ${fullMem.cubeId}`,
      ];

      if (fullMem.key) lines.push(`**Key**: ${fullMem.key}`);
      if (fullMem.tags.length > 0) lines.push(`**Tags**: ${fullMem.tags.join(", ")}`);
      if (fullMem.createdAt) lines.push(`**Created**: ${fullMem.createdAt}`);

      lines.push("", "### Content", "", fullMem.content);

      if (fullMem.background) {
        lines.push("", "### Background", "", fullMem.background);
      }

      if (fullMem.relations && fullMem.relations.length > 0) {
        lines.push("", "### Relations", "");
        for (const rel of fullMem.relations.slice(0, 5)) {
          lines.push(`- ${JSON.stringify(rel)}`);
        }
      }

      return [{ type: "text", text: lines.join("\n") }];
    } else {
      return [{ type: "text", text: notFoundText(memoryId) }];
    }
  } else if (result.data) {
    const errMsg = String((result.data as Record<string, unknown>).message ?? "Unknown error");
    if (errMsg.toLowerCase().includes("not found") || result.status === 404) {
      return [{ type: "text", text: notFoundText(memoryId) }];
    }
    return apiErrorResponse("Get", errMsg);
  } else {
    return apiErrorResponse("Get", `HTTP ${result.status}`);
  }
}

function notFoundText(memoryId: string): string {
  return [
    `❌ Memory not found: \`${memoryId}\``,
    "",
    "💡 **Tips**:",
    "- Verify the ID is correct (copy from memos_search results)",
    "- The memory may have been deleted",
    "- Try `memos_search` to find the memory again",
  ].join("\n");
}

// ============================================================================
// memos_get_stats
// ============================================================================

export async function handleMemosGetStats(arguments_: Record<string, unknown>): Promise<TextContent[]> {
  const cubeId = getCubeIdFromArgs(arguments_);

  const [regSuccess, regError] = await ensureCubeRegistered(cubeId);
  if (!regSuccess) return cubeRegistrationError(cubeId, regError);

  const result = await apiCallWithRetry(
    "GET",
    `${MEMOS_URL}/memories`,
    cubeId,
    { params: { user_id: MEMOS_USER, mem_cube_id: cubeId } },
    ensureCubeRegistered
  );

  if (result.success && result.data) {
    const [stats, total] = computeMemoryStats((result.data as Record<string, unknown>).data as SearchData ?? {});

    if (total === 0) {
      return [{ type: "text", text: `No memories found in cube '${cubeId}'.` }];
    }

    const typeIcons: Record<string, string> = {
      BUGFIX: "🐛", ERROR_PATTERN: "🔴", DECISION: "📋",
      GOTCHA: "⚠️", CODE_PATTERN: "📝", CONFIG: "⚙️",
      FEATURE: "✨", MILESTONE: "🎯", PROGRESS: "📊", INFERRED: "🔗",
    };

    const result_lines = [`## 📊 Memory Stats: ${cubeId}`, `Total Memories: **${total}**`, ""];

    for (const [mtype, count] of Object.entries(stats).sort((a, b) => b[1] - a[1])) {
      const percentage = (count / total * 100).toFixed(1);
      const icon = typeIcons[mtype] ?? "📌";
      result_lines.push(`- ${icon} **${mtype}**: ${count} (${percentage}%)`);
    }

    const inferredCount = stats.INFERRED ?? 0;
    const progressCount = stats.PROGRESS ?? 0;
    const userTyped = total - inferredCount - progressCount;

    if (inferredCount > 0) {
      result_lines.push("", "---", "", `ℹ️ **INFERRED** (${inferredCount} 条): 图数据库自动生成的因果推断节点，非用户保存，属正常现象。`);
    }

    if (total > 0 && progressCount / total > 0.5) {
      result_lines.push("", "---", "", `⚠️ **PROGRESS 占比偏高** (${progressCount}/${total}): 保存时建议显式指定类型:`);
      result_lines.push("   `BUGFIX` · `DECISION` · `MILESTONE` · `FEATURE` · `GOTCHA` · `CONFIG`");
    }

    if (userTyped > 0) {
      result_lines.push("", `✅ **用户标注记忆**: ${userTyped} 条 (${(userTyped / total * 100).toFixed(0)}%)`);
    }

    return [{ type: "text", text: result_lines.join("\n") }];
  } else if (result.data) {
    return apiErrorResponse("Stats", String((result.data as Record<string, unknown>).message ?? "Unknown error"));
  } else {
    return apiErrorResponse("Stats", `HTTP ${result.status}`);
  }
}

// ============================================================================
// Helpers
// ============================================================================

export function extractMemoriesFromData(data: SearchData): MemoryNode[] {
  const memories: MemoryNode[] = [];
  const textMems = data.text_mem ?? [];
  for (const cubeData of textMems) {
    const memData = cubeData.memories;
    if (memData && !Array.isArray(memData) && memData.nodes) {
      memories.push(...memData.nodes);
    } else if (Array.isArray(memData)) {
      memories.push(...(memData as MemoryNode[]));
    }
  }
  return memories;
}
